Use test groups for testing `error-stack`

indietyp opened the pull request:
## 🌟 What is the purpose of this PR?

error-stack recently gained new capabilities (#794, #793, #747), combined with the newly introduced extensive test suite, which has been implemented in #777, results in multiple new `feature` flags, which we need to test against. These circumstances have led to testing times up to ~5h, which - with new features - will only get exponentially longer.

This PR tries to find a compromise by introducing feature groups of features that are unlikely to influence each other. Discussion #877 outlines the worst-case math and potential improvements this change adds.

> **Note:** This change adds a new environmental flag, `CARGO_TEST_HACK_GROUPS`, which we append to the actual `cargo-hack` command. For the complete pre-release test, needs to set `CARGO_TEST_HACK_GROUPS = ""`, the env_script will pick it up and not emit any groups, running the full powerset.

> **Note:** I am unsure of the location for the env_scripts. Should this be in the main `Makefile`, then we only set `CARGO_TEST_HACK_GROUPS` in the `Makefile.toml` in the `error-stack` file?

> **Note:** I decided on an env_script approach because it got very unwieldy and confusing/unclear[^1] with such a long command (and makes the release test harder to realize). I think the current process will be better for maintainability.

## 🔗 Related links

- #877 

## 🚫 Blocked by

- https://github.com/sagiegurari/cargo-make/issues/685

## 🔍 What does this change?

* CI tests are run against a group of features
* pre-release tests are run against the full powerset of features

## 📜 Does this require a change to the docs?

- this is only internally used

## ⚠️ Known issues

## 🐾 Next steps

* change GitHub CI (I believe I do not have the permissions to do that) to be able to test on release

## 🛡 What tests cover this?
## ❓ How to test this?
## 📹 Demo

[^1]: German: unübersichtlich. I cannot find a 100% fitting word equivalent for the English language.

TimDiekmann reviewed (changes requested):
I appreciate the effort to parse this. However, I'm not really happy with this change as this complicates the CI flow a lot. Before we used `cargo-make`, we used shell scripts and it became unmaintainable, especially when the scripts are embedded inside of a declarative file.

Adding over 30 lines of code to only reduce the length of a parameter simply is not worth it, as there *will* be edge cases at some point.
Let's just put the `--group-features` into the flags and we are done

It's also not visible to me where we check if we are going to do a release, where does this happen?

indietyp commented:
Okay, I removed the script and added the `test-release` command (forgot that one). This now no longer uses any scripts and just makes use of environmental variables.

TimDiekmann reviewed (changes requested):
This is a lot clearer IMO, thanks! 🚀 

It's possible to use profiles in cargo-make. Currently, we run with the `development` profile and the `production` profile. I think it makes sense to overwrite only the environments like

```toml
[env]
CARGO_TEST_HACK_FLAGS_BASE = "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow"
CARGO_TEST_HACK_FLAGS = "${CARGO_TEST_HACK_FLAGS_BASE} --group-features anyhow,eyre --group-features std,hook --group-features spantrace,futures,small"

[env.release]
CARGO_TEST_HACK_FLAGS = "${CARGO_TEST_HACK_FLAGS_BASE}"
```

For the CI (`.github/workflows/rust.yml`) the `release` profile needs to be added to the `publish` job matrix. Also, the required tools have to be installed (see the `test` job). Then, `cargo make --profile ${{ matrix.profile }} test` can be added to the `publish` task.

It's possible to test the dry-run when changing the crate version. You can temporarily add a commit changing the version and revert this afterward.

indietyp commented:
Thanks, I'll do that!

indietyp commented:
I think I just found a bug in cargo-make? When I do:

```rust
[env]
CARGO_TEST_HACK_FLAGS_BASE = "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow"
CARGO_TEST_HACK_FLAGS = "${CARGO_TEST_HACK_FLAGS_BASE} --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small"
```

This fails, with `"cargo" "hack" "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small" "nextest" "run" "--cargo-profile" "dev"` as the issued command.

I changed it to

```rust
[env]
CARGO_TEST_HACK_FLAGS_BASE = "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow"

// snippet omitted

[tasks.test]
env = { CARGO_TEST_HACK_FLAGS = "${CARGO_TEST_HACK_FLAGS_BASE} --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small" }
```

This executes with: `"cargo" "hack" "--workspace" "--feature-powerset" "--exclude-features" "default" "--optional-deps" "eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small" "nextest" "run" "--cargo-profile" "dev"` as the issued command.

Strange.

indietyp commented:
Update: I cannot reproduce the error in a separate `Makefile.toml`. The investigation is continuing.

TimDiekmann commented:
I wonder, if we want to rename the profile to `"publish"` instead to avoid confusion with `"--release"`

TimDiekmann commented on packages/libs/error-stack/Makefile.toml:
```diff
@@ -5,6 +5,15 @@ CARGO_CLIPPY_HACK_FLAGS = "--workspace --feature-powerset --exclude-features def
 CARGO_MIRI_FLAGS = "--tests --lib"
 
+[tasks.test]
+env = { CARGO_TEST_HACK_FLAGS = "${CARGO_TEST_HACK_FLAGS} --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small" }
```
Currently reviewing https://github.com/hashintel/hash/pull/794, which requires a specific set of features for `cargo insta`. I think, that `std` and `hooks` shouldn't be grouped together. Generally, I think `std` shouldn't be grouped at all.

TimDiekmann replied:
complex-wise this shouldn't be a big impact as `hook` is implying `std`.

indietyp commented:
note: `publish` as profile name is not possible, as it is a reserved name

indietyp commented on .github/scripts/rust/Makefile.toml:
```diff
@@ -18,6 +18,10 @@ CARGO_DOC_HACK_FLAGS = "--workspace"
 CARGO_RUSTDOC_FLAGS = "--all-features -Zunstable-options -Zrustdoc-scrape-examples=examples -- -Zunstable-options"
 CARGO_RUSTDOC_HACK_FLAGS = "--workspace"
 CARGO_TEST_FLAGS = ""
+# This needs to be here, as error-stack uses it, and _BASE **needs** to be defined before
+# the variable using it has been defined. This is done per suggestion of
+# https://github.com/sagiegurari/cargo-make/issues/685
+CARGO_TEST_HACK_FLAGS_BASE = ""
```
This hotfix is no longer needed once [#688](https://github.com/sagiegurari/cargo-make/pull/688) lands. We could wait until then or just use the hotfix for now.

TimDiekmann replied:
I think we should keep the hot fix for now as this PR is preventing another workaround in #794, which is currently blocking us from releasing `error-stack@v0.2`

TimDiekmann commented on .github/workflows/rust.yml:
```diff
@@ -209,7 +209,10 @@ jobs:
       matrix:
         directory: ${{ fromJSON(needs.setup.outputs.publish) }}
         toolchain:
+          - stable
```
This will attempt to publish the library twice.
```suggestion
```

TimDiekmann reviewed (changes requested):
Looks good, only one issue which needs to be fixed.

I triggered a rerun for https://github.com/hashintel/hash/pull/892/commits/3e06c84f7e6f772d09e54739aa857fbf80ac5288 so a `publish` job was at least finished once.
